internal/liquids/nova/usage.go

```diff
@@ -0,0 +1,203 @@
+	if l.hasPooledResource.Get()[hwVersion][base] {
+		return liquid.ResourceName(fmt.Sprintf("hw_version_%s_%s", hwVersion, base))
+	}
+	return base
+}
+
+func (l *Logic) ScanUsage(ctx context.Context, projectUUID string, req liquid.ServiceUsageRequest, serviceInfo liquid.ServiceInfo) (liquid.ServiceUsageReport, error) {
+	var limitsData struct {
+		Limits struct {
+			Absolute struct {
+				MaxTotalCores         int64  `json:"maxTotalCores"`
+				MaxTotalInstances     int64  `json:"maxTotalInstances"`
+				MaxTotalRAMSize       int64  `json:"maxTotalRAMSize"`
+				MaxServerGroups       int64  `json:"maxServerGroups"`
+				MaxServerGroupMembers int64  `json:"maxServerGroupMembers"`
+				TotalCoresUsed        uint64 `json:"totalCoresUsed"`
+				TotalInstancesUsed    uint64 `json:"totalInstancesUsed"`
+				TotalRAMUsed          uint64 `json:"totalRAMUsed"`
+				TotalServerGroupsUsed uint64 `json:"totalServerGroupsUsed"`
+			} `json:"absolute"`
+			AbsolutePerFlavor map[string]struct {
+				MaxTotalInstances  int64  `json:"maxTotalInstances"`
+				TotalInstancesUsed uint64 `json:"totalInstancesUsed"`
+			} `json:"absolutePerFlavor"`
+			AbsolutePerHWVersion map[string]struct {
+				MaxTotalCores      int64  `json:"maxTotalCores"`
+				MaxTotalInstances  int64  `json:"maxTotalInstances"`
+				MaxTotalRAMSize    int64  `json:"maxTotalRAMSize"`
+				TotalCoresUsed     uint64 `json:"totalCoresUsed"`
+				TotalInstancesUsed uint64 `json:"totalInstancesUsed"`
+				TotalRAMUsed       uint64 `json:"totalRAMUsed"`
+			} `json:"absolutePerHwVersion"`
+		} `json:"limits"`
+	}
+	err := limits.Get(ctx, l.NovaV2, limits.GetOpts{TenantID: projectUUID}).ExtractInto(&limitsData)
+	if err != nil {
+		return liquid.ServiceUsageReport{}, err
+	}
+	absoluteLimits := limitsData.Limits.Absolute
+	var totalServerGroupMembersUsed uint64
+	if absoluteLimits.TotalServerGroupsUsed > 0 {
+		totalServerGroupMembersUsed, err = l.ServerGroupProber.GetMemberUsageForProject(ctx, projectUUID)
+		if err != nil {
+			return liquid.ServiceUsageReport{}, err
+		}
+	}
+
+	// initialize `Resources`
+	resources := map[liquid.ResourceName]*liquid.ResourceUsageReport{
+		"cores": {
+			Quota: &absoluteLimits.MaxTotalCores,
+			PerAZ: liquid.AZResourceUsageReport{Usage: absoluteLimits.TotalCoresUsed}.PrepareForBreakdownInto(req.AllAZs),
+		},
+		"instances": {
+			Quota: &absoluteLimits.MaxTotalInstances,
+			PerAZ: liquid.AZResourceUsageReport{Usage: absoluteLimits.TotalInstancesUsed}.PrepareForBreakdownInto(req.AllAZs),
+		},
+		"ram": {
+			Quota: &absoluteLimits.MaxTotalRAMSize,
+			PerAZ: liquid.AZResourceUsageReport{Usage: absoluteLimits.TotalRAMUsed}.PrepareForBreakdownInto(req.AllAZs),
+		},
+		"server_groups": {
+			Quota: &absoluteLimits.MaxServerGroups,
+			PerAZ: map[liquid.AvailabilityZone]*liquid.AZResourceUsageReport{liquid.AvailabilityZoneAny: {Usage: absoluteLimits.TotalServerGroupsUsed}},
```

Comment:
The liquid library has a convenience constructor for this:
```suggestion
			PerAZ: liquid.InAnyAZ(liquid.AZResourceUsageReport{Usage: absoluteLimits.TotalServerGroupsUsed}),
```
Same below for `server_group_members`.